### PR TITLE
changes for 0.5 and some small fixes

### PR DESCRIPTION
**Pull request by daywalker90**
## What does this PR do?
Fixes several type mismatches
Updated methods to non-deprecated ones in cln 23.05
adopt .save() changes from lnd.py to cln.py
introduces a 100h long fixed invoice expiry, because it gets dirty if you try to accept a htlc for an expired invoice in cln
adopt general changes to follow_send_payment and lookup_invoice_status
fix wrong field used in double_check_htlc_is_settled

### CODE REVIEW

**Review by Reckless-Satoshi — COMMENTED** [review]
First thing I tried, withdraw rewards, worked flawlessly! Then, I was "almost" capable of finishing an order! :D :rocket:

I believe there might be something happening with the final step of the order. Precisely with `LNNode.double_check_htlc_is_settled()`, I guess it returned `False` or errored even though the invoice had been settled. On a first click to "confirm fiat received", it did settle the escrow, but then it just returned before unlocking the bonds and triggering the payout to the buyer. On a second (and any subsequent click) I get:
```
_InactiveRpcError at /api/order/
<_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "Hodl-Invoice is in wrong state: `settled`. Payment_hash: bc3d430c7acb7ef4882d7184c081bc7e9e2b62dd2140bcb9df95f29720efd15d"
	debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2023-05-18T12:14:04.98625096+00:00", grpc_status:13, grpc_message:"Hodl-Invoice is in wrong state: `settled`. Payment_hash: bc3d430c7acb7ef4882d7184c081bc7e9e2b62dd2140bcb9df95f29720efd15d"}"
>
```
Seems to relate to the invoice being already settled.

File: api/lightning/cln.py [hunk 1]
```diff
@@ -458,60 +457,46 @@ def pay_invoice(cls, lnpayment):
...
             failure_reason = cls.payment_failure_context[status_code]
             lnpayment.failure_reason = status_code  # or failure_reason ?
             lnpayment.status = LNPayment.Status.FAILRO
-            lnpayment.save()
+            lnpayment.save(update_fields=["failure_reason", "status"])
             return False, failure_reason
 
     @classmethod
-    def follow_send_payment(cls, hash):
+    def follow_send_payment(cls, lnpayment, fee_limit_sat, timeout_seconds):
```
[the remark this answers is not in the document]
Comment (daywalker90) [reply api/lightning/cln.py:487]: what is a "not initiated" invoice ?

**Comment by Reckless-Satoshi** [comment]
> added commit for lnd.py to main:node.py parity

Thank you for catching these

**Comment by daywalker90** [comment]
- some more small fixes
- a little work on follow_send_payment to make it fit better for cln and add the invoice-expired-still-pending check
- add send_keysend method (no self payments)

**Comment by Reckless-Satoshi** [comment]
**A small summary of the state**
The RoboSats coordinator is now fully functional with CLN as backend using the hodlvoice plugin. 

Most tests needed to check the correct operation during the order pipeline have been done and are successful. It does, however, need more stressing and testing as there are still edge cases we have not covered.

Specifically, we need to test the cases where CLN is down while robosats makes requests (e.g. confirm fiat sent) and those cases where the hodl invoice expires or is closed to expiration while orders are ongoing.

**But...**
We are indeed going to be reckless and merge as is into Main! We can run more tests in testnet and keep committing fixes into Main. 

This huge task, adding support to CLN, is rewarded with 2M Sats #196 . To be honest, I can barely believe you have made it! Given the amount of hops and barriers in the way, including, building the CLN plugin... woah. Please, take some beers with friends to celebrate this!

Please, submit an invoice for 2M Sats with a long expiration time (+24h) from a proxy node id (not your own public node pubkey).

**Q:**
We will have to mantain our own CLN Dockerfile to include the hodl-grpc-plugin for the time being. Do you know if the hodlvoice gRPC is posed to be merged to ElementsProject/lightning repository? And if so, will the official core-lightning image builds (https://hub.docker.com/r/elementsproject/lightningd) eventually include it? If this is the case, we can just pull from them and always have an up-to-date CLN. Otherwise, we might want to publish in Dockerhub our own hodl-enabled CLN images, so it is trivial to stay up-to-date with CLN for robosats coordinators.

**Comment by daywalker90** [comment]
Thank you for the compliment @Reckless-Satoshi it was quite fun to do!

I'll reiterate it here again for documentation puroposes:
If an invoice is close to expiry (60s before) the htlcs will all be returned to the sender and the invoice will have the Hodlstate `Canceled` and the regular invoice lookup with cln will be `UNPAID` and shortly after `EXPIRED`.
This is probably also the biggest difference in behaviour to the lnd hodl-invoices, which is why we had to put a big expiry on all hodl-invoices in robosats.

With regards to you question: I'll ask if a merge is possible, but i kind of doubt it. I maintain a branch of cln in my repo that has the plugin merged if that helps: https://github.com/daywalker90/lightning/tree/lightning

PS: Thank you for helping me test this and the plugin!

lnbc20m1pjxk3nupp5mzj9hygvp9jaw0jkf4vtd7tq2f5f0q0dk0tffs49g4a0f8ukpu2qdqu2askcmr9wssx7e3q2dshgmmndp5scqzzsxqyz5vqsp5n27fw7hytth4pfrcq8e2k5je8nm8tewrcdt99l68szlahr9f0xss9qyyssqmr9xcpyru3ft77kfke50tyf5zvj2zksrsn8s36gcnhda8kr4kdmrdtv2yuvqzzndwsjvattw0hj69yvx4yuytjuph66ehff866sm4yqqe4ctfg

**Comment by daywalker90** [comment]
Thank you!

**Comment by Reckless-Satoshi** [comment]
> With regards to you question: I'll ask if a merge is possible, but i kind of doubt it. I maintain a branch of cln in my repo that has the plugin merged if that helps: https://github.com/daywalker90/lightning/tree/lightning

Will the the most up-to-date version be maintained in the `hodlvoice` branch or `lightning` branch? Our Dockerfile currently builds CLN form ElementsProject/lightning and the grpc plugin from daywalker90/lightning#hodlvoice .

What is the process or limitations to get this branch merged this with lightning/cln-grpc in ElementsProjects/lightning main? I believe most services would strictly prefer using hodl invoices even if for tiny hodling times, as simplified returns are top UX compared to asking the user to submit a refund invoice. Therefore this plugin adds some must have for the main cln-grpc.

**Comment by daywalker90** [comment]
If you mean most up-to-date version of the plugin then both, lightning contains the whole cln repo + plugin, which i use for rebasing and merging with cln master, hodlvoice is just the plugin with my own readme. You can keep using the hodlvoice branch.

I have not had the opportunity to ask for a possibility to merge. I'll keep you posted on that.
